src/main/java/org/apache/sysds/runtime/instructions/fed/MultiReturnParameterizedBuiltinFEDInstruction.java

```diff
@@ -0,0 +1,215 @@
+				FederatedResponse federatedResponse = pair.getValue().get();
+				if(federatedResponse.isSuccessful()) {
+					FrameBlock fb = (FrameBlock) federatedResponse.getData()[0];
+					combineRecodeMaps(combinedRecodeMaps, fb, range.getBeginDimsInt()[1]);
+				}
+			}
+		}
+		catch(InterruptedException | ExecutionException e) {
+			throw new DMLRuntimeException("Federated meta frame creation failed: " + e.getMessage());
+		}
+
+		// construct a single meta frameblock out of the multiple HashMaps with the recodings
+		FrameBlock meta = frameBlockFromRecodeMaps(combinedRecodeMaps);
+
+		// actually encode the frame block and construct an encoded matrix block at worker
+		List<Pair<Map.Entry<FederatedRange, FederatedData>, Future<FederatedResponse>>> encodedFutures = new ArrayList<>();
+		for(Map.Entry<FederatedRange, FederatedData> entry : fedMapping.entrySet()) {
+			FederatedRange fedRange = entry.getKey();
+			int columnStart = (int) fedRange.getBeginDims()[1];
+			int columnEnd = (int) fedRange.getEndDims()[1];
+			
+			// Slice out relevant meta part
+			// range is inclusive
+			FrameBlock slicedMeta = meta.slice(0, meta.getNumRows() - 1, columnStart, columnEnd - 1, null);
+			
+			Future<FederatedResponse> response = entry.getValue().executeFederatedOperation(new FederatedRequest(
+				FederatedRequest.FedMethod.FRAME_ENCODE, slicedMeta, spec, columnStart + 1), true);
+			encodedFutures.add(new ImmutablePair<>(entry, response));
+		}
+		
+		// construct a federated matrix with the encoded data
+		MatrixObject transformedMat = ec.getMatrixObject(getOutput(0));
+		transformedMat.getDataCharacteristics().set(fin.getDataCharacteristics());
+		Map<FederatedRange, FederatedData> transformedFedMapping = new HashMap<>();
+		try {
+			for(Pair<Map.Entry<FederatedRange, FederatedData>, Future<FederatedResponse>> data : encodedFutures) {
+				FederatedResponse federatedResponse = data.getValue().get();
+				if(federatedResponse.isSuccessful()) {
+					FederatedRange federatedRange = data.getKey().getKey();
+					FederatedData federatedData = data.getKey().getValue();
+					long varId = (long) federatedResponse.getData()[0];
+					
+					transformedFedMapping.put(federatedRange, new FederatedData(federatedData, varId));
+				}
+			}
+		}
+		catch(InterruptedException | ExecutionException e) {
+			throw new DMLRuntimeException("Federated transform apply failed: " + e.getMessage());
+		}
+		// set the federated mapping for the matrix
+		transformedMat.setFedMapping(transformedFedMapping);
+
+		// release input and outputs
+		ec.setFrameOutput(getOutput(1).getName(), meta);
+	}
+
+	private FrameBlock frameBlockFromRecodeMaps(Map<String, Long>[] combinedRecodeMaps) {
+		int rows = 0;
+		for(Map<String, Long> map : combinedRecodeMaps) {
+			if(map != null) {
+				rows = Integer.max(rows, map.size());
+			}
+		}
+		FrameBlock fb = new FrameBlock(combinedRecodeMaps.length, Types.ValueType.STRING);
+		fb.ensureAllocatedColumns(rows);
```

Comment:
There is no constructor for rows and columns, actually this loop anyway was a bit hard to develop, maybe because I wasn't as familiar with `Frame`s, or because the interface has a few pitfalls that I had problems with.